codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt

```diff
@@ -302,19 +302,44 @@ class KotlinJmespathExpressionVisitor(
         return flatMappingBlock(expression.right, left.identifier, left.shape, left.projected)
     }
 
+    private fun projection(expression: ProjectionExpression, parentName: String): VisitedExpression {
+        val left = when (expression.left) {
+            is SliceExpression -> slice(expression.left as SliceExpression, parentName)
+            else -> expression.left.accept(this)
+        }
+        requireNotNull(left.shape) { "projection is operating on nothing" }
+        return flatMappingBlock(expression.right, left.identifier, left.shape, left.projected)
+    }
+
     override fun visitSlice(expression: SliceExpression): VisitedExpression {
         throw CodegenException("SliceExpression is unsupported")
     }
 
-    override fun visitSubexpression(expression: Subexpression): VisitedExpression {
-        val leftName = expression.left.accept(this).identifier
+    private fun slice(expression: SliceExpression, parentName: String): VisitedExpression {
+        val startIndex = if (expression.start.isEmpty()) {
+            "0"
+        } else {
+            if (expression.start.asInt < 0) "$parentName.size${expression.start.asInt}" else expression.start.asInt
+        }
 
-        return when (val right = expression.right) {
-            is FieldExpression -> subfield(right, leftName)
-            is IndexExpression -> index(right, leftName)
-            is Subexpression -> subexpression(right, leftName)
-            else -> throw CodegenException("Subexpression type $right is unsupported")
+        val stopIndex = if (expression.stop.isEmpty()) {
+            "$parentName.size"
+        } else {
+            if (expression.stop.asInt < 0) "$parentName.size${expression.stop.asInt}" else expression.stop.asInt
         }
+
+        val subListName = addTempVar("subList", "$parentName?.subList($startIndex, $stopIndex)")
+        val stepListName = addTempVar("stepList", "$parentName?.toMutableList()")
+        writer.write("$stepListName?.clear()")
```

Comment:
**question**: Why clear the list instead of creating a new one `addTempVar("sliceList", "mutableListOf()")`

Better yet [slice](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/slice.html#kotlin.collections$slice(kotlin.Array((kotlin.collections.slice.T)),%20kotlin.collections.Iterable((kotlin.Int)))/indices) has an overload that takes an iterable and uses the values at the given indices.

```kotlin
val sliceList = addTempVar("sliceList", "$parentName?.slice($startIndex..$stopIndex step ${expression.step})
```

[Example](https://pl.kotl.in/a42StyoE4)

Reply:
Yeah, idk why I ended up thinking kotlin didn't have a slice function. It's using it now though